### PR TITLE
codeintel: Fix duplicate package providers

### LINKED ISSUE #25691: codeintel: Fix reference counts for indexes that re-declare packages
This can be seen most aggressively with lsif-node's output of sourcegraph/sourcegraph. When resolving xrepo code intelligence queries, we'll look to find what packages declare a given moniker.

In the case of sg/sg, client/{shared,branded,wildcard,etc} all declare version 0.0.0 in package.json. This means that every commit re-declares the same package, so that on xrepo results, we'll check basically every upload known for that repository for references.

This also has an effect with the reference count on each upload record (used for data retention). Currently, Cloud is unable to remove any sg/sg index because it's incorrectly referenced by _some_ future commit.

We need to look into how to ensure that there's a maximum number of uploads that can declare the same package version _for the same repository_.

### PR DESCRIPTION
**Pull request by efritz**
**TL;DR**: Fixes #25691. There was a bug with the previous way we linked what uploads refer to each other via the relationships formed in the `lsif_packages` and `lsif_references` table. This bug caused repositories that re-declare the same package versions on different commits (which turns out to be very common in practice) to link to _every_ use of that package. Instead of having one canonical package providing a particular library, _the majority_ of commits to the providing repository may claim to provide it. The bug has been squashed.

This PR makes a few changes:

- Consolidates the logic of the methods `UpdateNumReferences` (that would update the refcount of uploads by looking at what depends on _them_) and `UpdateDependencyNumReferences` (that would increment or decrement the refcount of an upload's dependencies, depending on whether the upload was being inserted or deleted) into `UpdaterReferenceCounts`, which handles the update for the entire set.
- Deprecates the column `num_references` column in favor of `reference_count`. This is a _flood the world_ solution, but makes a lot of sense since we can just abandon the old one in place and recalculate the num references from scratch. We already have an OOB migration that we can use to backfill the values (we just switch the references within the migrator over to the new column and it's M A G I C).
- Update all queries over `lsif_packages` to ensure that we're only looking at the **canonical** upload provider for a package (the oldest upload with the same scheme, name, version, indexer, and repository root). This includes the reference count queries themselves. The queries should now be well-documented.

### CODE REVIEW

**Comment by Numbers88s** [comment]
LGTM 🚀  Happy to approve it when it's ready to review. Anything missing?

File: enterprise/internal/codeintel/stores/dbstore/migration/reference_count.go [hunk 1]
```diff
@@ -24,7 +24,8 @@ func NewReferenceCountMigrator(store *dbstore.Store, batchSize int) oobmigration
 
 // Progress returns the ratio between the number of upload records that have been
```
Comment (eseliger) [inline enterprise/internal/codeintel/stores/dbstore/migration/reference_count.go:25]: What kind of dark magic is this repurposing an existing migrator 👀 😁

**Review by Strum355 — APPROVED** [review]
uploads.go make me go :gohno:

File: enterprise/internal/codeintel/stores/dbstore/uploads.go [hunk 2]
```diff
@@ -731,7 +791,13 @@ func (s *Store) HardDeleteUploadByID(ctx context.Context, ids ...int) (err error
...
 const hardDeleteUploadByIDQuery = `
 -- source: enterprise/internal/codeintel/stores/dbstore/uploads.go:HardDeleteUploadByID
-DELETE FROM lsif_uploads WHERE id IN (%s)
+WITH locked_uploads AS (
+	SELECT u.id
+	FROM lsif_uploads u
+	WHERE u.id IN (%s)
+	ORDER BY u.id FOR UPDATE
+)
+DELETE FROM lsif_uploads WHERE id IN (SELECT id FROM locked_uploads)
```
Comment (eseliger) [inline enterprise/internal/codeintel/stores/dbstore/uploads.go:800]: Is a FOR UPDATE necessary for deletes within a transaction?

File: enterprise/internal/codeintel/stores/dbstore/uploads.go [hunk 3]
```diff
@@ -896,112 +962,222 @@ func (s *Store) updateUploadRetention(ctx context.Context, protectedIDs, expired
...
+
 	idQueries := make([]*sqlf.Query, 0, len(ids))
 	for _, id := range ids {
 		idQueries = append(idQueries, sqlf.Sprintf("%s", id))
 	}
+	idListQuery := sqlf.Join(idQueries, ", ")
+
+	excludeCondition := sqlf.Sprintf("TRUE")
+	if dependencyUpdateType == DependencyReferenceCountUpdateTypeRemove {
+		excludeCondition = sqlf.Sprintf("u.id NOT IN (%s)", idListQuery)
```
Comment (eseliger) [inline enterprise/internal/codeintel/stores/dbstore/uploads.go:1021]: idea: Recently found this, reduces arguments sent and variety of versions of this SQL query ending up in query insights.

```suggestion
		excludeCondition = sqlf.Sprintf("u.id NOT = ANY (%s)", pq.Array(ids))
```

File: enterprise/internal/codeintel/stores/dbstore/uploads.go [hunk 4]
```diff
@@ -896,112 +962,222 @@ func (s *Store) updateUploadRetention(ctx context.Context, protectedIDs, expired
...
+			-- in this set, not just the ones that were recently inserted or deleted.
+			(p.scheme, p.name, p.version) IN (
+				SELECT p.scheme, p.name, p.version
+				FROM packages_defined_by_target_uploads p
+			)
+		) AND
+
+		-- Don't match deleted uploads. We may be dealing with uploads still in the
+		-- processing state, though, so we explicitly allow those here.
+		u.state IN ('completed', 'processing') AND
```
Comment (eseliger) [inline enterprise/internal/codeintel/stores/dbstore/uploads.go:1085]: not sure it's a valid state, what about errored? It'd eventually become processing again

**Review by chrismwendt — APPROVED** [review]
LGTM, nothing obvious jumps out to me, also :sqlno:

File: enterprise/internal/codeintel/stores/dbstore/xrepo.go [hunk 5]
```diff
@@ -61,8 +86,25 @@ SELECT
...
+FROM lsif_dumps_with_repository_name u
+WHERE u.id IN (SELECT id FROM canonical_uploads)
+`
+
+// packageRankingQueryFragment uses `lsif_uploads u` JOIN `lsif_packages p` to return a rank
+// for each row grouped by package and source code location and ordered by the associated Git
+// commit date.
+const packageRankingQueryFragment = `
+rank() OVER (
+	PARTITION BY
```
Comment (eseliger) [inline enterprise/internal/codeintel/stores/dbstore/xrepo.go:98]: did you steal that from my worker fairness spike? 👅

**Review by eseliger — APPROVED** [review]
After this: DB server disk 📉

File: enterprise/internal/codeintel/stores/dbstore/uploads.go (hunk 2 above)
Comment (efritz) [reply enterprise/internal/codeintel/stores/dbstore/uploads.go:800]: No, and I feel like this is a really fuzzy area of what I know is true and what I think might be something database gods tend to like.

I had seen a drop in lock contention/deadlock frequency on deletes when I ordered the IDs deterministically for delete operations. I suspect that's because the order that the locks were taken were in the order of the `IN (...)` values, which happened to cause no deadlocks when they were sorted.

This just replaces an explicit sort in the app code with an explicit sort in the database. I don't think this will perform any differently but it makes it a bit more consistent with other code that updates the same rows.

Happy to discuss more.

File: enterprise/internal/codeintel/stores/dbstore/uploads.go (hunk 4 above)
Comment (efritz) [reply enterprise/internal/codeintel/stores/dbstore/uploads.go:1085]: errored uploads won't have any data in the `lsif_packages` or `lsif_references` tables.

File: enterprise/internal/codeintel/stores/dbstore/xrepo.go (hunk 5 above)
Comment (efritz) [reply enterprise/internal/codeintel/stores/dbstore/xrepo.go:98]: 👀

**Comment by efritz** [comment]
@eseliger 

> After this: DB server disk 📉

Let's hope in a good way 😬

File: enterprise/internal/codeintel/stores/dbstore/uploads.go (hunk 3 above)
Comment (efritz) [reply enterprise/internal/codeintel/stores/dbstore/uploads.go:1021]: This isn't actually equivalent 🤔. A unit test failed when I committed this change.

Comment (eseliger) [reply enterprise/internal/codeintel/stores/dbstore/uploads.go:1021]: that scares me, I did this before for other parts of our code 👀 Actually, did you commit _just_ this? I think it needs some more than that, like removing the variables created before so they aren't unused. (And actually, I think it needs to be `!= ANY` thinking about it again..)

Comment (efritz) [reply enterprise/internal/codeintel/stores/dbstore/uploads.go:1021]: Wait, I got the syntax way wrong. It's `NOT (val = ANY(list))`, not `val != ANY(list)`.
